core/storage/ipfs/merkledag/impl/node_impl.hpp

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef FILECOIN_STORAGE_IPFS_MERKLEDAG_NODE_IMPL_HPP
+#define FILECOIN_STORAGE_IPFS_MERKLEDAG_NODE_IMPL_HPP
+
+#include <map>
+#include <memory>
+#include <string_view>
+#include <vector>
+
+#include <boost/optional.hpp>
+#include "primitives/cid/cid.hpp"
+#include "storage/ipfs/merkledag/impl/link_impl.hpp"
+#include "storage/ipfs/merkledag/impl/pb_node_encoder.hpp"
+#include "storage/ipfs/merkledag/node.hpp"
+
+namespace fc::storage::ipfs::merkledag {
+  class NodeImpl : public Node {
+   public:
+
+    size_t size() const override;
+
+    void assign(common::Buffer input) override;
+
+    const common::Buffer &content() const override;
+
+    outcome::result<void> addChild(const std::string &name,
+                                   std::shared_ptr<const Node> node) override;
+
+    outcome::result<std::reference_wrapper<const Link>> getLink(
+        const std::string &name) const override;
+
+    void removeLink(const std::string &name) override;
+
+    const CID &getCID() const override;
+
+    const common::Buffer &getRawBytes() const override;
+
+    void addLink(const Link &link) override;
+
+    std::vector<std::reference_wrapper<const Link>> getLinks() const override;
+
+    static std::shared_ptr<Node> createFromString(const std::string &content);
+
+    static outcome::result<std::shared_ptr<Node>> createFromRawBytes(
+        gsl::span<const uint8_t> input);
+
+   private:
+    mutable boost::optional<CID> cid_{};
```

Comment:
:'-(

Reply:
We should support general interface _fc::storage::ipfs::Block_, which instances often will be transferred by constant reference or pointer, so `Block::getCID()` can't be non-const. On the other hand - we can't recalculate CID every time was added new children link or modified content, because it's a large overhead. So, _mutable_ is very useful here.